### ISSUE TITLE
IMB data validation errors on main branch

### ISSUE THREAD

**Issue by darrylabbate** [open-mpi/ompi#10854]
Thank you for taking the time to submit an issue!

### What version of Open MPI are you using? (e.g., v3.0.5, v4.0.2, git branch name and hash, etc.)

`main` @ f42993c

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)

Download from https://www.open-mpi.org/nightly/main/

Configure flags:

```
CFLAGS=-pipe --enable-picky --enable-debug --without-verbs --with-ofi=/opt/amazon/efa/ --enable-mpi1-compatibility
```

### If you are building/installing from a git clone, please copy-n-paste the output from `git submodule status`.

N/A

### Please describe the system on which you are running

* Operating system/version: Amazon Linux 2
* Computer hardware: x86_64
* Network type: EFA

-----------------------------

## Details of the problem

We are seeing data validation errors for some IMB tests. Note that we (AWS EFA team) are actively working on this.

See also: [MTT result](https://mtt.open-mpi.org/index.php?limit=&wrap=&trial=&enable_drilldowns=&yaxis_scale=&xaxis_scale=&hide_subtitle=&split_graphs=&remote_go=&do_cookies=&phase=test_run&text_start_timestamp=2022-09-23+18%3A32%3A03+-+2022-09-24+18%3A32%3A03&text_platform_hardware=%5Ex86_64%24&show_platform_hardware=show&text_os_name=%5ELinux%24&show_os_name=show&text_mpi_name=%5Eompi-nightly-main%24&show_mpi_name=show&text_mpi_version=%5Emain-202209240241-f42993c%24&show_mpi_version=show&text_suite_name=all&show_suite_name=show&text_test_name=all&show_test_name=hide&text_np=all&show_np=show&text_full_command=&show_full_command=show&text_http_username=%5Eamazon%24&show_http_username=show&text_local_username=all&show_local_username=hide&text_platform_name=%5Eaws-amazonlinux-slurm-efa-installer%24&show_platform_name=show&click=Detail&phase=test_run&test_result=_run_f&text_compute_cluster_id=&text_os_version=&show_os_version=&text_platform_type=&show_platform_type=&text_submit_id=&text_hostname=&show_hostname=&text_mpi_get_id=&text_mpi_install_compiler_id=&text_mpi_install_configure_id=&text_mpi_install_id=&text_compiler_name=&show_compiler_name=&text_compiler_version=&show_compiler_version=&text_vpath_mode=&show_vpath_mode=&text_endian=&show_endian=&text_bitness=&show_bitness=&text_configure_arguments=&text_exit_value=&show_exit_value=&text_exit_signal=&show_exit_signal=&text_duration=&show_duration=&text_client_serial=&show_client_serial=&text_result_message=&text_result_stdout=&text_result_stderr=&text_environment=&text_description=&text_test_build_id=&text_test_build_compiler_id=&text_test_run_id=&text_test_name_id=&text_test_run_command_id=&text_test_suite_id=&text_performance_id=&text_launcher=&show_launcher=&text_resource_mgr=&show_resource_mgr=&text_network=&show_network=&text_parameters=&show_parameters=&lastgo=summary)

```
#-----------------------------------------------------------------------------
# Benchmarking Allgatherv 
# #processes = 576 
#-----------------------------------------------------------------------------
       #bytes #repetitions  t_min[usec]  t_max[usec]  t_avg[usec]      defects
            0          200         0.24         0.28         0.24         0.00
64: Error Allgatherv,size = 576,sample #0
Process 64: Got invalid buffer: 
Buffer entry: 0.000000 
pos: 1
Process 64: Expected    buffer: 
Buffer entry: 0.000000 
0: Error Allgatherv,size = 576,sample #0
Process 0: Got invalid buffer: 
Buffer entry: 0.000000 
32: Error Allgatherv,size = 576,sample #0
Process 32: Got invalid buffer: 
Buffer entry: 0.000000 
pos: 1
pos: 1
Process 0: Expected    buffer: 
Buffer entry: 0.000000 
Process 32: Expected    buffer: 
Buffer entry: 0.000000 
512: Error Allgatherv,size = 576,sample #0
Process 512: Got invalid buffer: 
Buffer entry: 0.000000 
pos: 1
Process 512: Expected    buffer: 
Buffer entry: 0.000000 
560: Error Allgatherv,size = 576,sample #0
Process 560: Got invalid buffer: 
Buffer entry: 0.000000 
pos: 1
Process 560: Expected    buffer: 
Buffer entry: 0.000000 
544: Error Allgatherv,size = 576,sample #0
Process 544: Got invalid buffer: 
Buffer entry: 0.000000 
pos: 1
Process 544: Expected    buffer: 
Buffer entry: 0.000000 
48: Error Allgatherv,size = 576,sample #0
Process 48: Got invalid buffer: 
Buffer entry: 0.000000 
pos: 1
Process 48: Expected    buffer: 
Buffer entry: 0.000000 
496: Error Allgatherv,size = 576,sample #0
Process 496: Got invalid buffer: 
Buffer entry: 0.000000 
pos: 1
Process 496: Expected    buffer: 
256: Error Allgatherv,size = 576,sample #0
Process 256: Got invalid buffer: 
Buffer entry: 0.000000 
pos: 1
Process 256: Expected    buffer: 
```

```
-----------------------------------------------------------------------------
# Benchmarking Alltoall 
# #processes = 24 
# ( 120 additional processes waiting in MPI_Barrier)
#-----------------------------------------------------------------------------
       #bytes #repetitions  t_min[usec]  t_max[usec]  t_avg[usec]      defects
            0            1         0.05         0.16         0.10         0.00
            1            1        76.13       106.14        86.63         0.00
            2            1        78.59       125.47        98.34         0.00
21: Error Alltoall,size = 96,sample #0
20: Error Alltoall,size = 96,sample #0
Process 20: Got invalid buffer: 
Buffer entry: 4.100000 
pos: 0
Process 20: Expected    buffer: 
Buffer entry: 20.100000 
19: Error Alltoall,size = 96,sample #0
```

```
#-----------------------------------------------------------------------------
# Benchmarking Reduce_scatter_block 
# #processes = 6 
# ( 138 additional processes waiting in MPI_Barrier)
#-----------------------------------------------------------------------------
       #bytes #repetitions  t_min[usec]  t_max[usec]  t_avg[usec]      defects
            0            1         0.06         0.09         0.07         0.00
            4            1        12.85        24.73        17.33         0.00
1: Error Reduce_scatter_block,size = 8,sample #0
5: Error Reduce_scatter_block,size = 8,sample #0
Process 5: Got invalid buffer: 
Buffer entry: 28.100000 
pos: 0
Process 5: Expected    buffer: 
Buffer entry: 62.099998
```

**Comment by wzamazon** [comment]
I will be working on it. Open the issue for awareness.

**Comment by jsquyres** [comment]
@wzamazon Is this an Open MPI core problem? Or something specific to EFA?  Does this affect any of the release branches?

**Comment by wckzhang** [comment]
I'm looking into it. It doesn't affect any release branches. This looks to be caused by one of the mtl/pml cm changes I did in the accelerator framework, doesn't affect ob1 as far as I can tell. I'm going to look into the mtl datatype packing code.

**Comment by wckzhang** [comment]
So I found the issue, in pml_cm.h there was a line added to call opal_convertor_prepare_for_send which shouldn't have been introduced as it does more than desired. Instead, I replace that check with a direct call to opal_convertor_accelerator_init, which only does the buffer detection. I'm guessing somewhere in OPAL_CONVERTOR_PREPARE, the convertor fields were overwritten.

**Comment by wckzhang** [comment]
So the specific field that is being overwritten is: convertor.pBaseBuf   = (unsigned char*)buf + datatype->super.true_lb;. This is set as buf instead. Just need to restore this field.

**Comment by wckzhang** [comment]
Updated the PR, should work now

**Comment by darrylabbate** [comment]
Closing; fixed in #10864
